homeassistant/components/recorder/core.py

```diff
@@ -577,14 +577,19 @@ def run(self) -> None:
 
         self.hass.add_job(self.async_connection_success)
 
-        # If shutdown happened before Home Assistant finished starting
-        if self._wait_startup_or_shutdown() is SHUTDOWN_TASK:
-            self.migration_in_progress = False
-            # Make sure we cleanly close the run if
-            # we restart before startup finishes
-            self._shutdown()
-            self.hass.add_job(self.async_set_recorder_ready)
-            return
+        if self.migration_is_live or schema_is_current:
+            # If the migrate is live or the schema is current, we need to
+            # wait for startup to complete. If its not live, we need to continue
+            # on.
+            self.hass.add_job(self.async_set_db_ready)
```

Comment:
Is calling this intended before checking if we're shutting down (and then calling it again)?

Reply:
Yes, thats how it was before.  We can't await the future here because its a deadlock because the recorder won't finish async_setup until the db is ready 

https://github.com/home-assistant/core/blob/148f96351052b0a4ba31e7d15dad16d7639e1ceb/homeassistant/components/recorder/__init__.py#L167